src/protocols/bdx/BdxMessages.cpp

```diff
@@ -0,0 +1,456 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements utility methods for working with some complex BDX messages.
+ */
+
+#include <protocols/bdx/BdxMessages.h>
+
+#include <support/BufBound.h>
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+
+#include <limits>
+
+namespace {
+constexpr uint8_t kVersionMask = 0x0F;
+} // namespace
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+void TransferInit::Pack(BufBound & aBuffer) const
+{
+    uint8_t proposedTransferCtl = 0;
+    bool widerange = (StartOffset > std::numeric_limits<uint32_t>::max()) || (MaxLength > std::numeric_limits<uint32_t>::max());
+
+    proposedTransferCtl |= Version & kVersionMask;
+    proposedTransferCtl = proposedTransferCtl | TransferCtlOptions.Raw();
+
+    BitFlags<uint8_t, RangeControlFlags> rangeCtlFlags;
+    rangeCtlFlags.Set(kDefLen, MaxLength > 0);
+    rangeCtlFlags.Set(kStartOffset, StartOffset > 0);
+    rangeCtlFlags.Set(kWiderange, widerange);
+
+    aBuffer.Put(proposedTransferCtl);
+    aBuffer.Put(rangeCtlFlags.Raw());
+    aBuffer.PutLE16(MaxBlockSize);
+
+    if (StartOffset > 0)
+    {
+        if (widerange)
+        {
+            aBuffer.PutLE64(StartOffset);
+        }
+        else
+        {
+            aBuffer.PutLE32(static_cast<uint32_t>(StartOffset));
+        }
+    }
+
+    if (MaxLength > 0)
+    {
+        if (widerange)
+        {
+            aBuffer.PutLE64(MaxLength);
+        }
+        else
+        {
+            aBuffer.PutLE32(static_cast<uint32_t>(MaxLength));
+        }
+    }
+
+    aBuffer.PutLE16(FileDesLength);
+    if (FileDesignator)
```

Comment:
nit: `(FileDesignator != nullptr)`. It is best if only actual bools are used without comparator.

Reply:
Ack